packages/loopring_v3.1/contracts/iface/IExchangeModule.sol

```diff
@@ -0,0 +1,66 @@
+/*
+
+  Copyright 2017 Loopring Project Ltd (Loopring Foundation).
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+pragma solidity ^0.5.11;
+
+import "./IVerificationKeyProvider.sol";
+
+
+/// @title IExchangeModule
+/// @author Brecht Devos - <brecht@loopring.org>
+contract IExchangeModule is IVerificationKeyProvider
```

Comment:
Do we really need to extend IVerificationKeyProvider? Inside `IAbstractModule` we already have a IVerificationKeyProvider as member.

Reply:
Currently for every block we store the exchange module that committed it. When verifying the block we get the verification keys for the block by getting the verification key directly from the exchange module. 

We could also pass in the `IVerificationKeyProvider` in `commitBlock` and store that address in the block instead of the exchange modules's address. That would also be totally fine. I ended up not doing that for subjective reasons:
- I kind of liked that we simply store the exchange module that committed the block inside the block itself. This gives a more direct mapping between module instance and blocks and could maybe come in handy for other functionality.
- Because the exchange module is stored this data could be compressed to a shorter ID instead of the complete 20 bytes address. This could end up saving an SSTORE which would save 20,000 gas/block commit.

Either way is fine for me really. :)

Reply:
I'm wondering if it is possible in the future that some IExchangeModules encapsulate logic that don't use ZKP at all...  and these blocks can be finalized at commitment....

Reply:
If it doesn't use ZKP then they don't need to commit a block, so the update can be applied directly to the exchange. But there's also not a lot you can do completely on-chain in the exchange itself I think, anything interesting needs to modify the Merkle tree which practically needs a circuit. Something like withdrawing from the Merkle tree in withdrawal could be done in an exchange module, but that's possible with the current interface, the module just provides some dummy verification keys that will never be used anyway (no blocks are ever committed for that).